sd/endpointer.go

```diff
@@ -0,0 +1,75 @@
+package sd
+
+import (
+	"io"
+
+	"github.com/go-kit/kit/endpoint"
+	"github.com/go-kit/kit/log"
+	iEndpoint "github.com/go-kit/kit/sd/internal/endpoint"
+)
+
+// Endpointer listens to a service discovery system and yields a set of
+// identical endpoints on demand. An error indicates a problem with connectivity
+// to the service discovery system, or within the system itself; an Endpointer
+// may yield no endpoints without error.
+type Endpointer interface {
+	Endpoints() ([]endpoint.Endpoint, error)
```

Comment:
> I agree that it's important in theory, but this is not how the current implementations behave

Yeah, that's a problem, and my bad :(

> If an implementation uses push model both internally and in the Instancer (via internal/instance.Cache), how do you see it returning an error from the pull model's Endpoints() method? I.e. which error would it be, the last one received via push (such as lost connection to sd backend)?

If the subscription to the Consul server is broken, then for the duration of the break, I would expect Instancer to return some sd.UnavailableError, perhaps with a Cause or Reason error field with the specific error returned by the consul client library. Likewise if (say) 3 DNS SRV polls in a row fail, I would expect to get exactly the same sd.UnavailableError.

Is this enough to go on?